Description: Apply changes to sage/libs/gap/util.pyx
 These changes happened between sage 8.4 and 8.5.beta6 and applying them
 helps to apply the large gap 4.10 patch cleanly.

From e273d65435d4cab288b26ff14dd1d3ccc8469b7b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Bissey?= <frp.bissey@gmail.com>
Date: Fri, 9 Nov 2018 09:15:12 +1300
Subject: Replace filter for compatibility with python3

---
 src/sage/libs/gap/util.pyx | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

From 842c6dfc15bd04c2e40441fb7b2c23608f210b9f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Bissey?= <frp.bissey@gmail.com>
Date: Fri, 9 Nov 2018 11:44:58 +1300
Subject: Figured out a way of doctesting gap_root alternate code

---
 src/sage/libs/gap/util.pyx | 8 ++++++++
 1 file changed, 8 insertions(+)

From 06b7daefa71636190b166d896bb0dd221a179cf7 Mon Sep 17 00:00:00 2001
From: Timo Kaufmann <timokau@zoho.com>
Date: Sun, 25 Nov 2018 14:13:43 +0100
Subject: Fix gap_root doctest when GAP_ROOT_DIR is reset

---
 src/sage/libs/gap/util.pyx | 36 ++++++++++++++++++++++--------------
 1 file changed, 22 insertions(+), 14 deletions(-)

From 7a661907c1db86447516304ce8c45f99ce7d3510 Mon Sep 17 00:00:00 2001
From: Timo Kaufmann <timokau@zoho.com>
Date: Mon, 26 Nov 2018 14:24:06 +0100
Subject: Improve the error message on invalid GAP_ROOT_DIR

---
 src/sage/libs/gap/util.pyx | 15 +++++++++------
 1 file changed, 9 insertions(+), 6 deletions(-)

--- a/sage/src/sage/libs/gap/util.pyx
+++ b/sage/src/sage/libs/gap/util.pyx
@@ -14,6 +14,8 @@
 
 from __future__ import print_function, absolute_import
 
+import os.path
+
 from cpython.exc cimport PyErr_SetObject
 from cpython.object cimport Py_LT, Py_LE, Py_EQ, Py_NE, Py_GT, Py_GE
 from cysignals.signals cimport sig_on, sig_off, sig_error
@@ -155,6 +157,25 @@
 ### Initialization of libGAP ###############################################
 ############################################################################
 
+def _guess_gap_root():
+    """
+    Used as a fallback to determine gapdir if if GAP_ROOT_DIR is undefined or
+    pointing to the wrong location.
+
+    EXAMPLES::
+
+        sage: from sage.libs.gap.util import _guess_gap_root
+        sage: _guess_gap_root()
+        The gap-4.5.5.spkg (or later) seems to be not installed!
+        ...
+    """
+    print('The gap-4.5.5.spkg (or later) seems to be not installed!')
+    gap_sh = open(os.path.join(SAGE_LOCAL, 'bin', 'gap')).read().splitlines()
+    gapdir = next(dir for dir in gap_sh if dir.strip().startswith('GAP_DIR'))
+    gapdir = gapdir.split('"')[1]
+    gapdir = gapdir.replace('$SAGE_LOCAL', SAGE_LOCAL)
+    return gapdir
+
 def gap_root():
     """
     Find the location of the GAP root install which is stored in the gap
@@ -166,15 +187,13 @@
         sage: gap_root()   # random output
         '/home/vbraun/opt/sage-5.3.rc0/local/gap/latest'
     """
-    import os.path
-    if os.path.exists(GAP_ROOT_DIR):
-        return GAP_ROOT_DIR
-    print('The gap-4.5.5.spkg (or later) seems to be not installed!')
-    gap_sh = open(os.path.join(SAGE_LOCAL, 'bin', 'gap')).read().splitlines()
-    gapdir = filter(lambda dir:dir.strip().startswith('GAP_DIR'), gap_sh)[0]
-    gapdir = gapdir.split('"')[1]
-    gapdir = gapdir.replace('$SAGE_LOCAL', SAGE_LOCAL)
-    return gapdir
+    try:
+        if os.path.exists(GAP_ROOT_DIR):
+            return GAP_ROOT_DIR
+    except TypeError:
+        raise RuntimeError('The GAP_ROOT_DIR environment variable is set to an invalid path: "{}"'.format(GAP_ROOT_DIR))
+
+    return _guess_gap_root()
 
 
 # To ensure that we call initialize_libgap only once.
